src/v/net/connection_rate.cc

```diff
@@ -0,0 +1,124 @@
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "net/connection_rate.h"
+
+#include "seastar/core/coroutine.hh"
+#include "ssx/future-util.h"
+#include "vassert.h"
+
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/sleep.hh>
+#include <seastar/net/inet_address.hh>
+
+#include <algorithm>
+#include <chrono>
+
+namespace net {
+
+connection_rate::connection_rate(
+  const connection_rate_info& rate_info, ss::gate& connection_gate) noexcept
+  : _general_rate(rate_info.max_connection_rate)
+  , _connection_gate(connection_gate) {
+    for (const auto& [ip, max_rate] : rate_info.ovverides) {
+        auto [_, res] = _ovverides.emplace(
+          ip, connection_rate_counter(max_rate));
+        vassert(
+          res, "Can not add ovveride for ip: {}, max_rate: {}", ip, max_rate);
+    }
+}
+
+ss::future<> connection_rate::lock(ss::sstring ip) {
+    auto& sem = find_sem(ip);
+    update_tokens_if_needed(sem);
+    co_await sem.rate.wait(_max_wait_time, 1);
+    spawn_updating_fiber_if_needed(ip, sem);
+}
+
+connection_rate_counter& connection_rate::find_sem(const ss::sstring& ip) {
+    auto ovverides_it = _ovverides.find(ip);
+    if (ovverides_it != _ovverides.end()) {
+        return ovverides_it->second;
+    }
+
+    return _general_rate;
+}
+
+void connection_rate::stop() {
+    _general_rate.rate.broken();
+    for (auto& [_, sem] : _ovverides) {
+        sem.rate.broken();
+    }
+}
+
+void connection_rate::update_tokens_if_needed(
+  connection_rate_counter& rate_counter) {
+    int64_t max_tokens_for_update = rate_counter.max_tokens
+                                    - rate_counter.rate.current();
+
+    if (max_tokens_for_update == 0) {
+        return;
+    }
+
+    auto now = ss::lowres_clock::now();
+    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(
+      now - rate_counter.last_update_time);
+
+    int64_t tokens_for_update = duration.count()
+                                / rate_counter.one_token_time.count();
+
+    if (tokens_for_update == 0) {
+        return;
+    }
+
+    rate_counter.last_update_time = now;
+
+    int64_t new_tokens_count = std::min(
+      tokens_for_update, max_tokens_for_update);
+
+    rate_counter.rate.signal(new_tokens_count);
+}
+
+void connection_rate::spawn_updating_fiber_if_needed(
+  const ss::sstring& ip, connection_rate_counter& rate_counter) {
+    if (rate_counter.rate.current() == 0) {
+        // Should spawn fiber to update tokens on next second
+        ssx::spawn_with_gate(
+          _connection_gate,
+          [this, ip, sleep_time = rate_counter.one_token_time]() mutable {
+              return ss::sleep(ss::lowres_clock::duration(sleep_time))
+                .then([this, ip] {
+                    auto& sem = find_sem(ip);
+                    update_tokens_force(sem);
+                });
+          });
+    }
+}
+
+void connection_rate::update_tokens_force(
```

Comment:
why not `reset_tokens` (i'm not sure what kind of tokens) - but don't get the 'force'. I think you mean reset